### ISSUE TITLE
[JENKINS-35881] API allows re-running of a completed run with the same commits

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#2848]
<p>In Scope</p>
<ul>
	<li>New rerun action in the rest API that can take the commits of a completed run and execute a new run with those exact commits.</li>
	<li>This should work for multi-branch pipeline jobs</li>
	<li>Ideally I'd like to see this work for other jobs too (such as Freestyle) if it isn't too much work</li>
	<li>If this does not work for all job types, the UI will need some way of testing if re-runs are supported (so we can show the button or not) – We should consider <a href="https://issues.jenkins.io/browse/JENKINS-35884" title="API for testing a entities capabilities for the UI" class="issue-link" data-issue-key="JENKINS-35884"><del>UX-350</del></a></li>
</ul>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay">jamesdumay</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35881" target="_blank">API allows re-running of a completed run with the same commits</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=imeredith">imeredith</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: blueocean-plugin
<li><b>label(s)</b>: api, blueocean-imported, japan, x-team
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-07-11T22:42:23+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 1
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>In Scope</p>
<ul>
	<li>New rerun action in the rest API that can take the commits of a completed run and execute a new run with those exact commits.</li>
	<li>This should work for multi-branch pipeline jobs</li>
	<li>Ideally I'd like to see this work for other jobs too (such as Freestyle) if it isn't too much work</li>
	<li>If this does not work for all job types, the UI will need some way of testing if re-runs are supported (so we can show the button or not) – We should consider <a href="https://issues.jenkins.io/browse/JENKINS-35884" title="API for testing a entities capabilities for the UI" class="issue-link" data-issue-key="JENKINS-35884"><del>UX-350</del></a></li>
</ul>
</pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35881?focusedId=261469&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261469">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=imeredith" class="user-hover" rel="imeredith">imeredith</a> is it still true that UX-158 blocks this work?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=imeredith" class="user-hover" rel="imeredith">imeredith</a> is it still true that UX-158 blocks this work?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35881?focusedId=261278&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261278">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>NOTE: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=imeredith" class="user-hover" rel="imeredith">imeredith</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a> if you use "Replay" feature as it is today, it will actaully run it with the same commit, so this is essentially the same as <a href="https://cloudbees.atlassian.net/browse/UX-416" class="external-link" target="_blank" rel="nofollow noopener">https://cloudbees.atlassian.net/browse/UX-416</a> - only in one case you provide an alternative Jenkinsfile content, in another case you don't change it - but same result. </p>

<p>Perhaps we should close this as a duplicate of <a href="https://cloudbees.atlassian.net/browse/UX-416" class="external-link" target="_blank" rel="nofollow noopener">https://cloudbees.atlassian.net/browse/UX-416</a> - and track it there? 
I don't think replay works with freestyle, or plain pipeline jobs, so that may be a limitation (in which case, all the api could do is re-run the job, not with a specific commit, and I would be ok with that limitation). </p>
</pre>
</details>

</ul>
</details>

<p>NOTE: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=imeredith" class="user-hover" rel="imeredith">imeredith</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a> if you use "Replay" feature as it is today, it will actaully run it with the same commit, so this is essentially the same as <a href="https://cloudbees.atlassian.net/browse/UX-416" class="external-link" target="_blank" rel="nofollow noopener">https://cloudbees.atlassian.net/browse/UX-416</a> - only in one case you provide an alternative Jenkinsfile content, in another case you don't change it - but same result. </p>

<p>Perhaps we should close this as a duplicate of <a href="https://cloudbees.atlassian.net/browse/UX-416" class="external-link" target="_blank" rel="nofollow noopener">https://cloudbees.atlassian.net/browse/UX-416</a> - and track it there? <br/>
I don't think replay works with freestyle, or plain pipeline jobs, so that may be a limitation (in which case, all the api could do is re-run the job, not with a specific commit, and I would be ok with that limitation). </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35881?focusedId=261311&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261311">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Nah keep <a href="https://issues.jenkins.io/browse/JENKINS-35839" title="Support Replay of a pipeline which allows a user to update a pipeline script before retrying" class="issue-link" data-issue-key="JENKINS-35839"><del>UX-416</del></a> open - its a different use case entirely. I've gone ahead and linked the two.</p>
</pre>
</details>

</ul>
</details>

<p>Nah keep <a href="https://issues.jenkins.io/browse/JENKINS-35839" title="Support Replay of a pipeline which allows a user to update a pipeline script before retrying" class="issue-link" data-issue-key="JENKINS-35839"><del>UX-416</del></a> open - its a different use case entirely. I've gone ahead and linked the two.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35881?focusedId=261341&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261341">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a> they are different, but could be quite close both in experience and implementation, but one could argue that when you go to replay+edit, you are in "fix up build config" mode, vs "re-run against this commit to see if it is environmental failure" - different frames of mind when you do it, agree. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a> they are different, but could be quite close both in experience and implementation, but one could argue that when you go to replay+edit, you are in "fix up build config" mode, vs "re-run against this commit to see if it is environmental failure" - different frames of mind when you do it, agree. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35881?focusedId=261264&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-261264">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mneale" class="user-hover" rel="mneale">mneale</a> I still think the modify+rerun is very important. For the sake of brevity I just want to split these problems out for the moment.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mneale" class="user-hover" rel="mneale">mneale</a> I still think the modify+rerun is very important. For the sake of brevity I just want to split these problems out for the moment.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>imeredith:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35881?focusedId=262695&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-262695">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Blocked by an issue in workflow plugin. PR submitted</p>
</pre>
</details>

</ul>
</details>

<p>Blocked by an issue in workflow plugin. PR submitted</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35881?focusedId=262706&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-262706">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Associated PR: <a href="https://github.com/jenkinsci/workflow-cps-plugin/pull/27" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/workflow-cps-plugin/pull/27</a></p>
</pre>
</details>

</ul>
</details>

<p>Associated PR: <a href="https://github.com/jenkinsci/workflow-cps-plugin/pull/27" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/workflow-cps-plugin/pull/27</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-35881?focusedId=263048&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263048">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=imeredith" class="user-hover" rel="imeredith">imeredith</a> workflow-cps 2.8 has been released with your changes.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=imeredith" class="user-hover" rel="imeredith">imeredith</a> workflow-cps 2.8 has been released with your changes.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `depends_on` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-36289%22">JENKINS-36289</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-35789%22">JENKINS-35789</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-35794%22">JENKINS-35794</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-35839%22">JENKINS-35839</a>]</i>
